std/experimental/allocator/building_blocks/bitmapped_block.d

```diff
@@ -1746,3 +1760,350 @@ private struct BitVector
+            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2,
+            2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
+            2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
+            4, 4, 4, 4, 4, 4, 4, 4, 5, 5, 5, 5, 6, 6, 7
+        ];
+
+        for (size_t i = 0; i < _control.length; i++)
+        {
+            ulong controlVal, newControlVal, bitIndex;
+            do
+            {
+                bitIndex = 0;
+                newControlVal = 0;
+                controlVal = atomicLoad(_control[i]);
+
+                // skip all control words which have all bits set
+                if (controlVal == ulong.max)
+                    break;
+
+                // fast lookup of first byte which has at least one zero bit
+                foreach (byteIndex; 0 .. 8)
+                {
+                    ulong mask = (0xFFUL << (8 * (7 - byteIndex)));
+                    if ((mask & controlVal) != mask)
+                    {
+                        ubyte byteVal = cast(ubyte) ((mask & controlVal) >> (8 * (7 - byteIndex)));
+                        bitIndex += firstZero[byteVal];
+                        newControlVal = controlVal | (1UL << (63 - bitIndex));
+                        break;
+                    }
+                    bitIndex += 8;
+                }
+            } while (!cas(&_control[i], controlVal, newControlVal));
+
+            auto blockIndex = bitIndex + 64 * i;
+            if (controlVal != ulong.max && blockIndex < _blocks)
+            {
+                size_t payloadBlockStart = cast(size_t) blockIndex * blockSize;
+                return cast(void[]) _payload[payloadBlockStart .. payloadBlockStart + s];
+            }
+        }
+
+        return null;
+    }
+
+    /**
+    Deallocates the given buffer `b`, by atomically setting the corresponding
+    bit to `0`. `b` must be valid, and cannot contain multiple adjacent `blocks`.
+    */
+    bool deallocate(void[] b)
+    {
+        import core.atomic : cas, atomicLoad;
+
+        if (b is null)
+            return true;
+
+        ulong controlVal, newControlVal;
+        auto blockIndex = (b.ptr - _payload.ptr) / blockSize;
+        auto controlIndex = blockIndex / 64;
+        auto bitIndex = blockIndex % 64;
+
+        do
+        {
+            controlVal = atomicLoad(_control[controlIndex]);
+            newControlVal = controlVal & ~(1 << (63 - bitIndex));
+        } while (!cas(&_control[controlIndex], controlVal, newControlVal));
+
+        return true;
+    }
+
+    /**
+    Expands in place a buffer previously allocated by `SharedBitmappedBlock`.
+    Expansion fails if the new length exceeds the block size.
+    */
+    bool expand(ref void[] b, immutable size_t delta)
+    {
+        if (delta == 0)
+            return true;
+
+        immutable newLength = delta + b.length;
+        if (b is null || newLength > blockSize)
+            return false;
+
+        b = b.ptr[0 .. newLength];
```

Comment:
Why `b.ptr`? `b = b[0 .. newLength]` should work too.
(Besides `.ptr` is unsafe...)

Reply:
if the newLength is greater than the previous length, it doesn't work without the ptr